cpp/src/arrow/array/data.h

```diff
@@ -180,25 +180,46 @@ struct ARROW_EXPORT ArrayData {
 
   std::shared_ptr<ArrayData> Copy() const { return std::make_shared<ArrayData>(*this); }
 
-  bool IsNull(int64_t i) const { return !IsValid(i); }
+  inline bool IsNull(int64_t i) const { return !IsValid(i); }
 
-  bool IsValid(int64_t i) const {
+  inline bool IsValid(int64_t i) const {
     if (buffers[0] != NULLPTR) {
       return bit_util::GetBit(buffers[0]->data(), i + offset);
     }
     const auto type = this->type->id();
     if (type == Type::SPARSE_UNION) {
       return !internal::IsNullSparseUnion(*this, i);
-    }
-    if (type == Type::DENSE_UNION) {
+    } else if (type == Type::DENSE_UNION) {
       return !internal::IsNullDenseUnion(*this, i);
-    }
-    if (type == Type::RUN_END_ENCODED) {
+    } else if (type == Type::RUN_END_ENCODED) {
       return !internal::IsNullRunEndEncoded(*this, i);
     }
     return null_count.load() != length;
   }
 
+  template <typename ArrowType>
+  bool IsNullFast(int64_t i) const {
+    return !IsValidFast<ArrowType>(i);
+  }
+
+  template <typename ArrowType>
+  bool IsValidFast(int64_t i) const {
+    if constexpr (ArrowType::type_id == Type::NA) {
+      return false;
+    } else if constexpr (ArrowType::type_id == Type::SPARSE_UNION) {
+      return !internal::IsNullSparseUnion(*this, i);
+    } else if constexpr (ArrowType::type_id == Type::DENSE_UNION) {
+      return !internal::IsNullDenseUnion(*this, i);
+    } else if constexpr (ArrowType::type_id == Type::RUN_END_ENCODED) {
+      return !internal::IsNullRunEndEncoded(*this, i);
+    } else {
+      if (buffers[0] != NULLPTR) {
+        return bit_util::GetBit(buffers[0]->data(), i + offset);
+      }
+      return null_count.load() != length;
```

Comment:
Maybe a dumb question: if the validity bitmap does not exist, shouldn't `null_count` be either 0 or -1 instead of other values?

Reply:
`NA` is a known type that has no validity bitmap and `null_count` set to `length`. So in theory, this is very possible.